File: Emulator/src/main/content/global/skill/fletching/items/bolts/GemBoltPulse.java
package content.global.skill.fletching.items.bolts;

import core.game.node.entity.player.Player;
import core.game.node.entity.skill.SkillPulse;
import core.game.node.entity.skill.Skills;
import core.game.node.item.Item;

/**
 * Represents the attaching of a gem bolt to a premade bolt.
 *
 * @author Ceikry
 */
public final class GemBoltPulse extends SkillPulse<Item> {

    /**
     * Represents the gem bolt being made.
     */
    private GemBolt bolt;

    /**
     * Represents the sets to make.
     */
    private int sets = 0;

    /**
     * Represents the ticks passed.
     */
    private int ticks;

    /**
     * Constructs a new {@code GemBoltPulse} {@code Object}.
     *
     * @param player the player.
     * @param node   the node.
     * @param bolt   the bolt
     * @param sets   the sets.
     */
    public GemBoltPulse(Player player, Item node, GemBolt bolt, int sets) {
        super(player, node);
        this.bolt = bolt;
        this.sets = sets;
    }

    @Override
    public boolean checkRequirements() {
        if (player.getSkills().getLevel(Skills.FLETCHING) < bolt.getLevel()) {
            player.getDialogueInterpreter().sendDialogue("You need a Fletching level of " + bolt.getLevel() + " or above to do that.");
            return false;
        }
        if (!player.getInventory().containsItem(new Item(bolt.getBase())) || !player.getInventory().containsItem(new Item(bolt.getTip()))) {
            return false;
        }
        if (!player.getInventory().hasSpaceFor(new Item(bolt.getProduct()))) {
            player.getDialogueInterpreter().sendDialogue("You do not have enough inventory space.");
            return false;
        }
        return true;
    }

    @Override
    public void animate() {
    }

    @Override
    public boolean reward() {
        if (++ticks % 3 != 0) {
            return false;
        }
        int baseAmount = player.getInventory().getAmount(bolt.getBase());
        int tipAmount = player.getInventory().getAmount(bolt.getTip());
        Item base = new Item(bolt.getBase());
        Item tip = new Item(bolt.getTip());
        Item product = new Item(bolt.getProduct());
        if (baseAmount >= 10 && tipAmount >= 10) {
            base.setAmount(10);
            tip.setAmount(10);
            product.setAmount(10);
        } else {
            int amount = Math.min(baseAmount, tipAmount);
            base.setAmount(amount);
            tip.setAmount(amount);
            product.setAmount(amount);
        }
        if (player.getInventory().remove(base, tip)) {
            player.getInventory().add(product);
            player.getSkills().addExperience(Skills.FLETCHING, bolt.getExperience() * product.getAmount(), true);
            player.getPacketDispatch().sendMessage(product.getAmount() == 1 ? "You attach the tip to the bolt." : "You fletch " + product.getAmount() + " bolts.");
        }
        sets--;
        return sets <= 0;
    }

}

File: Server/src/main/content/global/skill/fletching/items/bolts/GemBoltPulse.kt
package content.global.skill.fletching.items.bolts

import core.api.getStatLevel
import core.api.hasSpaceFor
import core.api.inInventory
import core.api.sendDialogue
import core.game.node.entity.player.Player
import core.game.node.entity.skill.SkillPulse
import core.game.node.entity.skill.Skills
import core.game.node.item.Item
import kotlin.math.min

/**
 * Represents the attaching of a gem bolt to a premade bolt.
 *
 * @author Ceikry
 */
class GemBoltPulse(player: Player?, node: Item?, private val bolt: GemBolt, sets: Int) : SkillPulse<Item?>(player, node) {
    /**
     * Represents the sets to make.
     */
    private var sets = 0

    /**
     * Represents the ticks passed.
     */
    private var ticks = 0

    init {
        this.sets = sets
    }

    override fun checkRequirements(): Boolean {
        if (getStatLevel(player, Skills.FLETCHING) < bolt.level) {
            sendDialogue(player, "You need a Fletching level of " + bolt.level + " or above to do that.")
            return false
        }
        if (!inInventory(player, bolt.base) || !inInventory(player, bolt.tip)) {
            return false
        }
        if (!hasSpaceFor(player, Item(bolt.product))) {
            sendDialogue(player, "You do not have enough inventory space.")
            return false
        }
        return true
    }

    override fun animate() {
    }

    override fun reward(): Boolean {
        if (++ticks % 3 != 0) {
            return false
        }
        val baseAmount = player.inventory.getAmount(bolt.base)
        val tipAmount = player.inventory.getAmount(bolt.tip)
        if (baseAmount <= 0 || tipAmount <= 0) {
            val materials = if (baseAmount <= 0) bolt.base else bolt.tip
            player.packetDispatch.sendMessage("You do not have any more $materials to fletch.")
            return true
        }
        val base = Item(bolt.base)
        val tip = Item(bolt.tip)
        val product = Item(bolt.product)
        if (baseAmount >= 10 && tipAmount >= 10) {
            base.amount = 10
            tip.amount = 10
            product.amount = 10
        } else {
            val amount = min(baseAmount.toDouble(), tipAmount.toDouble()).toInt()
            base.amount = amount
            tip.amount = amount
            product.amount = amount
        }
        if (player.inventory.remove(base, tip)) {
            player.inventory.add(product)
            player.getSkills().addExperience(Skills.FLETCHING, bolt.experience * product.amount, true)
            player.packetDispatch.sendMessage(if (product.amount == 1) "You attach the tip to the bolt." else "You fletch " + product.amount + " bolts.")
        }
        sets--
        return sets <= 0
    }
}
